_source/_posts/2022-07-20-express-prisma-tailwind.md

```diff
@@ -0,0 +1,665 @@
+
+### Complete the development of your express application
+
+Now that your database is ready, you can turn your attention to the core of your application.
+
+First, add a new file in the root of your application called `ensureLoggedIn.js` and add the following code to it.
+
+```javascript
+function ensureLoggedIn(req, res, next) {
+    if (req.isAuthenticated()) {
+        return next();
+    }
+    res.redirect('/login')
+}
+
+module.exports = ensureLoggedIn;
+```
+
+This little piece of middleware will make sure the user is authenticated.  If they aren't, you will redirect them to the `login` route, which you will configure to use Okta.  Otherwise, you will allow the user to reach the next screen.  
+
+Next, you can update your routes in the `routes` directory.  First, remove `users.js` as you won't be using that.  Next, replace the code in `index.js` with the following.
+
+```javascript
+var express = require('express');
+var router = express.Router();
+
+/* GET home page. */
+router.get('/', function(req, res, next) {
+  res.render('index', { title: 'The Workout Tracker - Home', isAuthenticated: req.isAuthenticated() });
+});
+
+module.exports = router;
+```
+
+Here you will pass a property for `isAuthenticated` so your layout page can properly display a login or logout button.  
+
+Next, add a file in the `routes` directory for `dashboard.js` with the following code.
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+
+const ensureLoggedIn = require("../ensureLoggedIn");
+
+/* GET home page. */
+router.get("/", ensureLoggedIn, async function (req, res) {
+  const username = req.user.username;
+
+  const dbUser = await prisma.user.findUnique({
+    where: {
+      username: username,
+    },
+    include: {
+      workoutLogs: true,
+    },
+  });
+
+  res.render("dashboard", {
+    title: "The Workout Tracker - Dashboard",
+    isAuthenticated: req.isAuthenticated(),
+    user: dbUser
+  });
+});
+
+module.exports = router;
+```
+
+There's a lot of magic in this route.  First, you are using the `ensureLoggedIn` middleware to protect this route.  Then you can extract the username from the request and look up the user in the database using the Prisma client.  Since the user has a property for its workout logs, you can pass the user as the model and parse the logs on the pug view you will create.  This will give the view access to the user's name as well.  Of course, you could query the `workoutLogs` table directly and include features such as pagination and searching if that better fit your workflow.  
```

Comment:
```suggestion
There's a lot of magic in this route.  First, you are using the `ensureLoggedIn` middleware to protect the route.  Then you can extract the username from the request and look up the user in the database using the Prisma client.  Since the user has a property for its workout logs, you can pass the user as the model and parse the logs on the pug view you will create.  This will give view access to the user's name as well.  Of course, you could query the `workoutLogs` table directly and include features such as pagination and search if this approach better fits your workflow.  
```
alignment of verb tense in closing sentence